File: content/courses/parallel-computing-introduction/codes/mpi_subarray.f90
program mpi_subarray
    use mpi_f08
    implicit none

    integer, parameter :: root = 0
    integer :: ierr, rank, size
    integer :: nrl, ncl
    integer :: sub_rows, sub_cols
    integer :: starts(2), subsizes(2), wsizes(2)
    integer :: i, j
    integer :: lb, ubr, ubc
    integer, allocatable :: w(:,:)
    type(MPI_Datatype)::sendtype,recvtype

    call MPI_Init()
    call MPI_Comm_rank(MPI_COMM_WORLD, rank)
    call MPI_Comm_size(MPI_COMM_WORLD, size)

    ! For simplicity, run with 2 processes
    if (size /= 2) then
        if (rank == root) print *, "Please run with 2 processes."
        call MPI_Finalize()
        stop
    end if

    ! Define the full array dimensions
    nrl = 6
    ncl = 8

    ! Define the subarray dimensions
    sub_rows = 3
    sub_cols = 4

    lb=0
    ubr=nrl-1
    ubc=ncl-1

    allocate(w(lb:ubr, lb:ubc))

    ! Fill the arrays.  All zeros on rank 1.
    if (rank == root) then
        do i = lb, ubr
            do j = lb, ubc
                w(i, j) = i * 100 + j
            end do
        end do
    else
        w=0
    end if

    ! Set up subarrays
    wsizes = [nrl, ncl]                 ! Full array size
    subsizes = [ sub_rows, sub_cols ]   ! Subarray size

    ! Send array
    starts   = [ 2, 3 ]                 ! Starting indices

    ! Subarray type requires counting from 0
    starts=starts-lb

    call MPI_Type_create_subarray(2, wsizes, subsizes, starts, &
                                  MPI_ORDER_FORTRAN, MPI_INTEGER, sendtype)
    call MPI_Type_commit(sendtype)

    ! Receive array
    starts=[ 3,2 ]-lb

    call MPI_Type_create_subarray(2, wsizes, subsizes, starts, &
                                  MPI_ORDER_FORTRAN, MPI_INTEGER, recvtype, ierr)
    call MPI_Type_commit(recvtype)

    ! Send subarray from root to rank 1
    if (rank == root) then
        print *, "Rank", rank, "sending array:"
        do i = lb, ubr
            write(*,'(*(i6.3))') w(i,:)
        end do
        call MPI_Send(w, 1, sendtype, 1, 0, MPI_COMM_WORLD)
    else
        call MPI_Recv(w, 1 , recvtype, root, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE)
        ! Print received subarray
        print *, "Rank", rank, "received subarray:"
        do i = lb, ubr
            write(*,'(*(i6.3))') w(i,:)
        end do
    end if

    call MPI_Type_free(sendtype)
    call MPI_Type_free(recvtype)
    call MPI_Finalize()

end program mpi_subarray

File: content/courses/parallel-computing-introduction/codes/mpi_subarray.cxx
#include <cstring>
#include <iostream> 
#include <iomanip> 
#include <mpi.h> 

using namespace std; 

// MPI subarray example

int main (int argc, char *argv[]) {

    int nprocs, rank;
    int tag=0;
    int root=0;

    //Initialize MPI
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    //Run with two processes only
    if ( nprocs != 2 ) {
        if ( rank==root) { 
            cout<<"Please run with two processes"<<endl; 
            MPI_Finalize();
            exit(1);
        }
    }

    int nrl=6;
    int ncl=8;

    int **w=new int*[nrl];
    int *wptr=new int[(nrl)*(ncl)];

    for (int i=0;i<nrl;++i,wptr+=ncl) {
       w[i] = wptr;
    }

    if (rank==root) {
        for ( int i = 0; i < nrl; i++ ) {
             for (int j = 0; j < ncl; j++ ) {
                 w[i][j] = i*100+j;
             }
        }
    }
    else {
        for ( int i = 0; i < nrl; i++ ) {
             for (int j = 0; j < ncl; j++ ) {
                 w[i][j] = 0;
             }
        }
    }

    //Set up subarrays
    //
    int ndims=2;
    int sizes[]={nrl,ncl};

    int starts[ndims]; 
    int subsizes[ndims];

    subsizes[0]=3; subsizes[1]=4;

    MPI_Datatype sendtype, recvtype;

    starts[0]=2; starts[1]=3;
    MPI_Type_create_subarray(ndims,sizes,subsizes,starts,MPI_ORDER_C,MPI_INT,&sendtype);
    MPI_Type_commit(&sendtype);

    starts[0]=3; starts[1]=2;
    MPI_Type_create_subarray(ndims,sizes,subsizes,starts,MPI_ORDER_C,MPI_INT,&recvtype);
    MPI_Type_commit(&recvtype);

    //Buffers set up

    // Send type from root to rank 1
    if ( rank==root ) {
        cout<<"Rank "<<rank<<" sending array:"<<endl;
        for (int i=0;i<nrl;i++) {
            for (int j=0;j<ncl;j++) {
                cout<<setfill('0')<<setw(3)<<w[i][j]<<" ";
            }
            cout<<endl;
        }
        MPI_Send(&w[0][0], 1, sendtype, 1, tag, MPI_COMM_WORLD);
    }
    else {
        MPI_Recv(&w[0][0], 1, recvtype, root, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
        cout<<" Rank "<<rank<<" received subarray: "<<endl;
        for (int i=0;i<nrl;i++) {
            for (int j=0;j<ncl;j++) {
                cout<<setfill('0')<<setw(3)<<w[i][j]<<" ";
            }
            cout<<endl;
        }
    }

    //Type_free not really necessary but good practice
    MPI_Type_free(&sendtype);
    MPI_Type_free(&recvtype);

    MPI_Finalize();

    return 0;

}
